dev/encoder/quadrature_encoder.cpp

```diff
@@ -0,0 +1,144 @@
+// that it can be interpreted as a 32-bit signed or unsigned value, and it will
+// work anyway.
+//
+// As an example, a two wheel robot being controlled at 100Hz, can use two
+// state machines to read the two encoders and in the main control loop it can
+// simply ask for the current encoder counts to get the absolute step count. It
+// can also subtract the values from the last sample to check how many steps
+// each wheel as done since the last sample period.
+//
+// One advantage of this approach is that it requires zero CPU time to keep the
+// encoder count updated and because of that it supports very high step rates.
+//
+
+// 374 pulses per revolution (from the description at https://www.dfrobot.com/product-1462.html)
+// 4x encoding bc quadrature (?)
+const float ROT_PER_TICK = 1.0 / (4 * 374.0);
+// TODO: update pulley ratio
+const float PULLEY_RATIO = 0.3185 / 1.528;
+const float DEG_PER_ROT = 360.0;
+
+class Encoder
+{
+public:
+    // Create an encoder. Reccommended NOT to use this class, use EncoderFactory::createEncoder()
+    // @param pinA the pin that encoder A channel is connected to, the B channel should connect to the next pin
+    // @param sm the state machine to keep track of the encoder, 0-3
+    // @param which pio
+    Encoder(uint pinA, uint sm, PIO pio, float ratio, bool addProgram = true)
+    {
+        this->pio = pio;
+        this->sm = sm;
+        this->ratio = ratio;
+
+        uint offset = 0;
+
+        // we don't really need to keep the offset, as this program must be loaded
+        // at offset 0
+        if (addProgram)
+            uint offset = pio_add_program(pio, &quadrature_encoder_program);
+
+        quadrature_encoder_program_init(pio, sm, offset, pinA, 0);
+    }
+
+    // updates the pos and velocity, call periodically.
+    // @param delta_time the time, in miliseconds, since last calling update
+    void update(int delta_time)
+    {
+        pos = quadrature_encoder_get_count(pio, sm) * ratio * DEG_PER_ROT;
+        velocity = ((prev_pos - pos) / delta_time) * 1000;
+        prev_pos = pos;
+    }
+
+    // get position of wheel in rotations. resets on init.
+    // update() must be called preiodically for this to be accurate
+    float get_pos()
+    {
+        return pos;
+    }
+
+    // get velocity of wheel in rotations per second.
+    // update() must be called preiodically for this to be accurate
+    float get_velocity()
+    {
+        return velocity;
+    }
+
+private:
+    float prev_pos, pos;
+    float velocity;
+    float ratio;
+    PIO pio;
+    uint sm;
+};
+
+class EncoderFactory
+{
+public:
+    // Create an encoder, automatically configuring the state machine and pio.
+    // @param pinA the A encoder channel, the B channel should be connected to the next pin
+    // @param ratio the ratio by which to multiply encoder outputs. ratio of 1 results in tick / sec
+    static Encoder createEncoder(uint pinA, float ratio)
```

Comment:
should ratio be optional (default value 1)?

Reply:
yeah